scripts/split_repo_migration/git-migrate-history.sh

```diff
@@ -0,0 +1,146 @@
+SOURCE_PATH=$3
+
+# path in the target repo to put the copied code
+TARGET_PATH=$4
+
+# comma-separated list of files/folders to skip
+IGNORE_FOLDERS=$5
+# keep these specific files that would otherwise be deleted by IGNORE_FOLDERS
+KEEP_FILES=$6
+
+# override the HEAD branch name for the migration PR
+BRANCH=$7
+
+if [[ ! -z "${UPDATE_SCRIPT}" ]]
+then
+  UPDATE_SCRIPT=$(realpath "${UPDATE_SCRIPT}")
+fi
+
+if [[ -z "${BRANCH}" ]]
+then
+  # default the branch name to be generated from the source repo name
+  BRANCH=$(basename ${SOURCE_REPO})-migration
+fi
+echo "here"
+export FILTER_BRANCH_SQUELCH_WARNING=1
+
+# create a working directory
+WORKDIR=$(mktemp -d -t code-migration-XXXXXXXXXX)
+echo "Created working directory: ${WORKDIR}"
+
+pushd "${WORKDIR}"  # cd into workdir
+
+echo "Cloning source repository: ${SOURCE_REPO}"
+git clone "git@github.com:${SOURCE_REPO}.git" source-repo
+
+pushd source-repo
+git remote remove origin
+
+# prune only files within the specified directory
+if [[ ! -z "${SOURCE_PATH}" ]]
+then
+  echo "Pruning commits only including path: ${SOURCE_PATH}"
+  git filter-branch \  # rewrites history... unsafe
+    --prune-empty \
+    --subdirectory-filter "${SOURCE_PATH}"
+fi
+
+if [[ ! -z "${IGNORE_FOLDERS}" ]]
+then
+  echo "Ignoring folder: ${IGNORE_FOLDERS}"
+  mkdir -p ${WORKDIR}/filtered-source
+  FOLDERS=$(echo ${IGNORE_FOLDERS} | tr "," " ")
+  # remove files/folders we don't want
+  FILTER="(rm -rf ${FOLDERS} || true)"
+  if [[ ! -z "${KEEP_FILES}" ]]
+  then
+    KEEP_FILES_SPACES=($(echo ${KEEP_FILES} | tr "," " "))
+    LAST_ELEMENT=$(( ${#KEEP_FILES_SPACES[@]} - 1 ))
+    KEEP_FILE_COMMANDS=""
+    for file in "${KEEP_FILES_SPACES[@]}"
+    do
+      if [[ $file == "${KEEP_FILES_SPACES[$LAST_ELEMENT]}" ]]
+      then
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true"
+      else 
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true; "
+      fi   
+    done
+    # restore files to keep, silence errors if the file doesn't exist
+    FILTER="${FILTER}; ${KEEP_FILE_COMMANDS}"
+  fi
+  git filter-branch \
+    --force \
+    --prune-empty \
+    --tree-filter "${FILTER}"
+fi
+
+# reorganize the filtered code into the desired target locations
+if [[ ! -z "${TARGET_PATH}" ]]
+then
+  echo "Moving files to destination path: ${TARGET_PATH}"
+  git filter-branch \
+    --force \
+    --prune-empty \
+    --tree-filter \
+      "shopt -s dotglob; mkdir -p ${WORKDIR}/migrated-source; mv * ${WORKDIR}/migrated-source; mkdir -p ${TARGET_PATH}; mv ${WORKDIR}/migrated-source/* ${TARGET_PATH}"
+fi
+
+# back to workdir
+popd
+
+# merge histories
+echo "Cloning target repository: ${SOURCE_REPO}"
+git clone "git@github.com:${TARGET_REPO}.git" target-repo
+pushd target-repo
+
+git remote add --fetch migration ../source-repo
+git checkout -b "${BRANCH}"
+git merge --allow-unrelated-histories migration/main --no-edit
+
+if [[ ! -z "${UPDATE_SCRIPT}" ]]
```

Comment:
LMK whether you want to commit before calling `$UPDATE_SCRIPT`, which would result in 2 commits per migration, or have everything in one commit.

If the latter, let me know whether you want to commit here; I'm currently committing at the end of `$UPDATE_SCRIPT`

Reply:
I think having two commits help with debugging and potential rollbacks.

Reply:
I'm realizing I wasn't thinking about things correctly and used the wrong language here, so let me try to be accurate now.

This script pushes all the split-repo's commits (git commits) to the monorepo branch `$BRANCH`. So we have hundreds of commits, not just one or two.

IIUC (correct me if I'm wrong!), we have two options:

(A) We could issue a PR at this point. IIUC, the PR would simply merge the `$BRANCH` with this whole imported history into the `main` branch. Then after that's merged, we run my postprocessing script on the `main` branch directly and issue a PR against the `main` branch. This means that between the two PRs, there's a period where the imported split-repo shows up as a package dir in the monorepo `main`, but hasn't been fully migrated because the post-processing script hasn't run.

or

(B) On this branch `$BRANCH`, after all these imported commits, we run my post-process script and add one final commit with those changes. Then we issue a single PR to merge the whole branch into `main`. When it is merged, the new package directory will have all the migration changes it needs; we won't ever be in a state where the repo dir appears migrated but isn't really. If we want to review just the post-processing changes, we merely look at the last commit in the PR.

Writing this out, I think option B would be cleaner. WDYT? Please let me know whether there are aspects I'm failing to consider.

Reply:
Talked about this offline - the path forward here is to apply an update script in batch on all split repos after they have been migrated. This results in two PRs.

Reply:
Removed update script.